pkg/auth/handler/sso/auth_handler.go

```
@@ -0,0 +1,295 @@
+//
+// curl http://localhost:3000/sso/<provider>/auth_handler?code=<code>&state=<state>
+//
+// For ux_mode is 'ios' or 'android',
+// it creates a 302 response, and Location points to:
+// myapp://user.skygear.io/sso/{provider}/auth_handler?result=
+//
+// Fox ux_mode is 'web_redirect',
+// it creates a 302 response, and Location points to: sso_callback_url
+// and set cookie in the response.
+//
+// For ux_mode is 'web_popup',
+// it will render a html page and set cookie in the response.
+//
+type AuthHandler struct {
+	TxContext               db.TxContext                `dependency:"TxContext"`
+	AuthContext             coreAuth.ContextGetter      `dependency:"AuthContextGetter"`
+	OAuthAuthProvider       oauth.Provider              `dependency:"OAuthAuthProvider"`
+	PasswordAuthProvider    password.Provider           `dependency:"PasswordAuthProvider"`
+	AuthInfoStore           authinfo.Store              `dependency:"AuthInfoStore"`
+	RoleStore               role.Store                  `dependency:"RoleStore"`
+	TokenStore              authtoken.Store             `dependency:"TokenStore"`
+	AuthHandlerHTMLProvider sso.AuthHandlerHTMLProvider `dependency:"AuthHandlerHTMLProvider"`
+	ProviderFactory         *sso.ProviderFactory        `dependency:"SSOProviderFactory"`
+	Provider                sso.Provider
+	SSOSetting              sso.Setting
+	ProviderName            string
+}
+
+func (h AuthHandler) DecodeRequest(request *http.Request) (handler.RequestPayload, error) {
+	payload := AuthRequestPayload{}
+	q := request.URL.Query()
+	payload.Code = q.Get("code")
+	payload.Scope = strings.Split(q.Get("scope"), " ")
+	payload.EncodedState = q.Get("state")
+
+	return payload, nil
+}
+
+func (h AuthHandler) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	// reference from APIHandlerToHandler
+	var err error
+
+	defer func() {
+		if err != nil {
+			response := handler.APIResponse{
+				Err: skyerr.MakeError(err),
+			}
+			handler.WriteResponse(rw, response)
```

I find there are multiple functions that write things to the `ResponseWriter` which makes the flow hard to follow.

And errors are ignored in `handleSessionResp` and `handleRedirectResp`.